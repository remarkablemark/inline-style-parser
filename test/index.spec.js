const cssParse = require('css/lib/parse');
const inlineStyleParser = require('..');
const { cases, errors, snapshots } = require('./data');

/**
 * @param {string} inlineStyle
 * @return {Array}
 */
function getDeclarations(inlineStyle) {
  const {
    stylesheet: { rules }
  } = cssParse(`a{${inlineStyle}}`);
  return rules[0].declarations;
}

/**
 * @param {Object[]} declarations
 * @return {Object[]}
 */
function removePosition(declarations) {
  return declarations.map(({ type, property, value }) => ({
    type,
    property,
    value
  }));
}

// compare output with `css.parse` declarations
describe('when compared to `css.parse`', () => {
  it.each(cases)('correctly parses "%s"', (style) => {
    expect(removePosition(inlineStyleParser(style))).toEqual(
      removePosition(getDeclarations(style))
    );
  });
});

// match snapshots
describe('inline-style-parser', () => {
  it.each(snapshots)('%s', (_name, style) => {
    expect(inlineStyleParser(style)).toMatchSnapshot();
  });
});

// errors
describe('error', () => {
  it.each(errors)('throws when argument is `%s`', (value, message) => {
    expect(() => inlineStyleParser(value)).toThrow(message);
  });
});

// one-off
describe('misc', () => {
  describe.each`
    style  | options             | expected
    ${':'} | ${undefined}        | ${[]}
    ${'a'} | ${{ silent: true }} | ${[]}
  `(
    'when style="$style" and options=`$options`',
    ({ style, options, expected }) => {
      it(`returns expected output \`${JSON.stringify(expected)}\``, () => {
        expect(inlineStyleParser(style, options)).toEqual(expected);
      });
    }
  );
});
