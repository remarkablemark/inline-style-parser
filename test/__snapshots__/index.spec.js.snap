// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`inline-style-parser correctly parses multiple declarations 1`] = `
[
  {
    "position": Position {
      "end": {
        "column": 90,
        "line": 2,
      },
      "source": undefined,
      "start": {
        "column": 7,
        "line": 2,
      },
    },
    "property": "background",
    "type": "declaration",
    "value": "-webkit-gradient(linear, left top, left bottom, from(white), to(black))",
  },
  {
    "position": Position {
      "end": {
        "column": 24,
        "line": 3,
      },
      "source": undefined,
      "start": {
        "column": 7,
        "line": 3,
      },
    },
    "property": "content",
    "type": "declaration",
    "value": "" "",
  },
  {
    "comment": " comment ",
    "position": Position {
      "end": {
        "column": 39,
        "line": 3,
      },
      "source": undefined,
      "start": {
        "column": 26,
        "line": 3,
      },
    },
    "type": "comment",
  },
  {
    "position": Position {
      "end": {
        "column": 14,
        "line": 4,
      },
      "source": undefined,
      "start": {
        "column": 7,
        "line": 4,
      },
    },
    "property": "foo",
    "type": "declaration",
    "value": "bar",
  },
  {
    "position": Position {
      "end": {
        "column": 5,
        "line": 5,
      },
      "source": undefined,
      "start": {
        "column": 15,
        "line": 4,
      },
    },
    "property": "-o-transition",
    "type": "declaration",
    "value": "all .5s",
  },
]
`;

exports[`inline-style-parser correctly parses single declaration 1`] = `
[
  {
    "position": Position {
      "end": {
        "column": 26,
        "line": 1,
      },
      "source": undefined,
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "property": "background-color",
    "type": "declaration",
    "value": "#C0FFEE",
  },
]
`;

exports[`inline-style-parser correctly parses when comment precedes colon 1`] = `
[
  {
    "position": Position {
      "end": {
        "column": 40,
        "line": 1,
      },
      "source": undefined,
      "start": {
        "column": 1,
        "line": 1,
      },
    },
    "property": "text-align",
    "type": "declaration",
    "value": "center",
  },
]
`;
